app/src/main/java/org/sagebionetworks/research/mpower/viewmodel/SurveyViewModel.kt

```diff
@@ -46,12 +47,17 @@ import org.threeten.bp.LocalDateTime
 /**
  * SurveyViewModel contains a live data object that queries for all non-excluded surveys unfinished and available now.
  */
-open class SurveyViewModel(app: Application): ScheduleViewModel(app) {
+open class SurveyViewModel(private var scheduleDao: ScheduledActivityEntityDao,
+        scheduleRepo: ScheduleRepository) : ScheduleViewModel(scheduleDao, scheduleRepo) {
+
+    class Factory @Inject constructor(private val scheduledActivityEntityDao: ScheduledActivityEntityDao,
```

Comment:
Dagger + Kotlin is relatively new, I found [this post](https://github.com/google/dagger/issues/900) interesting.

An inner class in Kotlin like this for a Factory doesn't seem correct to me, because it should probably be a companion object, possibly with @JvmStatic annotation, but I really don't have precedent to concretely say how it should be.

Reply:
One additional factor here is that to get a non-zero-arg-constructor `ViewModel` from the framework through `ViewModelProviders`, we need to instantiate a factory that implements `ViewModelProvider.Factory`. Is there a way to do that with companion objects? Would moving this to a top level class be better?

Reply:
Companion objects are an interesting concept, I've never really seen the equivalent in another language.  They can actually be whatever sub-class you want. 

```
companion object: ViewModelProvider.Factory {
        override fun <T : ViewModel?> create(modelClass: Class<T>): T {
       }
}
```

Reply:
However, I don't think you can do anything with constructors which it looks like is how dagger works. The "object" keyword is most like anonymous inner classes in Java which you also can't add constructors to.